### PR TITLE
[Console] Add CommandDefinition to make `list` lazy

### LINKED ISSUE #33804: [Console] Distinguish definition of a command from its execution
**Description**  
Currently a command is a single PHP class, which implements both: its definition and the code to execute it. This is perfectly fine for most use cases, but has downsides for more complicated ones.

For commands that implement business logic and makes heavy use of dependency injection, creating the command object can be very expensive.

This fact was greatly mitigated by introducing `\Symfony\Component\Console\CommandLoader\CommandLoaderInterface`. Using that, will lazily instantiate only the command object specified, not all command objects like before. 
However we still face this performance impact when listing all commands (`console list`), where again all command objects are built "just" to show their names and the description (access the definition).

Therefore I'd propose to *optionally* separate definition from execution, by allowing commands to implement an interface which could roughly look like (don't pick on the names or details it's just an outline):

```php
<?php
namespace Symfony\Component\Console\Command;

interface CommandDefinitionInterface
{
    /**
     * @return CommandDefinition Includes InputDefinition, name, description, help...
     */
    public function getDefinition(): CommandDefinition;

    /**
     * Could also be part of CommandDefinition instead
     * 
     * @return string Class name or service name, which implements CommandExecutorInterface or is closure
     */
    public function getExecutor(): string;
}

interface CommandExecutorInterface
{
    /**
     * Executes command logic.
     *
     * @return int|null null or 0 if everything went fine, or an error code
     */
    public function execute(InputInterface $input, OutputInterface $output);
}
```

The benefit of this (compared to `setCode`, which could be deprecated then) is, that `ContainerCommandLoader` could be made responsible for creating the executor instead of the command object having to have a dependency to a DI container.

Any thoughts? Thanks!

### PR DESCRIPTION
**Pull request by jderusse**
| Q             | A
| ------------- | ---
| Branch?       | 5.x
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | yes
| Tickets       | Fix #33804
| License       | MIT
| Doc PR        | -

This is an alternative to #39851, that use an `CommandDescription` to expose the command's metadata.

This PR is WIP to get your point of view before moving spending too much time on it.

TODO:
* [ ] add tests
* [ ] trigger exception when command implement the interface **and** override the methods `isEnabled`, `isHidden`, ... meaning
* [ ] add changelog
* [ ] migrate commands (with dependencies) to this interface

### CODE REVIEW

File: src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php [hunk 1]
```diff
@@ -1757,7 +1757,7 @@ private function dumpValue($value, bool $interpolate = true): string
 
                     $code = sprintf('return %s;', $code);
 
-                    return sprintf("function ()%s {\n            %s\n        }", $returnedType, $code);
+                    return sprintf("function () use (\$container) %s {\n            %s\n        }", $returnedType, $code);
```
Comment (jderusse) [inline src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php:1760]: This probably should be migrated to a dedicated PR + tests

The previous code trigger a `Notice: Undefined variable: container` exception  (#39851 should also fix this issue).
Here is the dumped code:
```
        return $container->privates['.console.command.cache_clear.lazy'] = new \Symfony\Component\Console\Command\LazyCommand(function () : \Symfony\Bundle\FrameworkBundle\Command\CacheClearCommand {
            return ($container->privates['console.command.cache_clear'] ?? $container->load('getConsole_Command_CacheClearService'));
        });
```

File: src/Symfony/Component/Console/Application.php [hunk 2]
```diff
@@ -489,9 +489,6 @@ public function add(Command $command)
             return null;
         }
 
-        // Will throw if the command is not correctly initialized.
-        $command->getDefinition();
```
Comment (jderusse) [inline src/Symfony/Component/Console/Application.php:493]: This break laziness (#39851 should also fix this issue).
I'm not sure why this check is performed here.

**Comment by OskarStark** [comment]
This looks very promising and less "magical" 👍🏻

**Comment by ro0NL** [comment]
`DescribableCommandInterface` should not be needed IMHO

with >=php8 we can argue attributes are superior either way, and thus we're targetting php7 only

It's really fast to meet the convention from  #39851 for the time being (eg. pre-php8 era): `$defaultName = 'foo|bar';` and done

File: src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php (hunk 1 above)
Comment (nicolas-grekas) [reply src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php:1760]: Depending on the way the container is dumped, the container should be referenced as `$container` or `$this`.

Look for `use ($container)` in the current `PhpDumper` to figure out how this should be done.

Also for sure, this is a bug fix and it should be submitted on a dedicated PR.

File: src/Symfony/Component/Console/Application.php (hunk 2 above)
Comment (nicolas-grekas) [reply src/Symfony/Component/Console/Application.php:493]: Related to https://github.com/symfony/symfony/pull/9244 apparently
Apparently, we need to think twice before removing as that might hurt DX.

**Comment by nicolas-grekas** [comment]
From a DX perspective, this requires ppl to implement a new static method to gain laziness, which is more code to write.

#39851 provides a descriptive approach instead, that paves the way to configuring the same thing using attributes on PHP 8. Doing so would be a minor implementation change on top of #39851.

What benefit are you expecting from this approach, vs #39851?

As of now, the proposed `CommandDescription` looks very much like a PHP 8 attribute in disguise.
If we agree on the target that would be to allow using an attribute for the purposed config settings, then I think #39851 is more aligned to this goal.

File: src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php (hunk 1 above)
Comment (jderusse) [reply src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php:1760]: see #39873

**Comment by jderusse** [comment]
> What benefit are you expecting from this approach

Cleaner, extensible (ie. allows enabled), self explanatory (`|foo|bar` is not obviously a hidden command).

> If we agree on the target that would be to allow using an attribute for the purposed config settings, then I think #39851 is more aligned to this goal.

I totally agree with this comment (and similar from @ro0NL), if we target php8 attributes, that would make sens to move with #39851 and make the migration smoother

**Comment by nicolas-grekas** [comment]
> Cleaner, extensible (ie. allows enabled), self explanatory (|foo|bar is not obviously a hidden command).

On my side, it looks like more code boilerplate for something that could be just plain descriptive.
About the "self-explanatory" part, it's a very minor thing, and the attribute will improve this anyway.

**Comment by jderusse** [comment]
Closing in favor of #39851 and a path to php8 attributes
